cryptoki-rustcrypto/src/ecdsa.rs

```diff
@@ -0,0 +1,176 @@
+    SignatureAlgorithmIdentifier,
+};
+use std::{convert::TryFrom, ops::Add};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Cryptoki error: {0}")]
+    Cryptoki(#[from] cryptoki::error::Error),
+
+    #[error("Private key missing attribute: {0}")]
+    MissingAttribute(AttributeType),
+
+    #[error("Elliptic curve error: {0}")]
+    Ecdsa(#[from] ecdsa::elliptic_curve::Error),
+}
+
+pub trait SignAlgorithm: PrimeCurve + CurveArithmetic + AssociatedOid + DigestPrimitive {
+    fn sign_mechanism() -> Mechanism<'static>;
+}
+
+impl SignAlgorithm for p256::NistP256 {
+    fn sign_mechanism() -> Mechanism<'static> {
+        Mechanism::Ecdsa
+    }
+}
+
+pub struct Signer<C: SignAlgorithm> {
+    session: Session,
+    _public_key: ObjectHandle,
+    private_key: ObjectHandle,
+    verifying_key: VerifyingKey<C>,
+}
+
+impl<C: SignAlgorithm> Signer<C>
+where
+    FieldBytesSize<C>: ModulusSize,
+    AffinePoint<C>: FromEncodedPoint<C> + ToEncodedPoint<C>,
+{
+    pub fn new(session: Session, label: &[u8]) -> Result<Self, Error> {
+        // First we'll lookup a private key with that label.
+        let template = vec![
+            Attribute::Token(true),
+            Attribute::Private(true),
+            Attribute::Label(label.to_vec()),
+            Attribute::Class(ObjectClass::PRIVATE_KEY),
+            Attribute::KeyType(KeyType::EC),
+            Attribute::EcParams(C::OID.to_der().unwrap()),
+            Attribute::Sign(true),
+        ];
+
+        let private_key = session.find_objects(&template)?.remove(0);
+        let attribute_pk = session.get_attributes(private_key, &[AttributeType::Id])?;
+
+        // Second we'll lookup a public key with the same label/ec params/ec point
+        let mut template = vec![
+            Attribute::Private(false),
+            Attribute::Label(label.to_vec()),
+            Attribute::Class(ObjectClass::PUBLIC_KEY),
+            Attribute::KeyType(KeyType::EC),
+            Attribute::EcParams(C::OID.to_der().unwrap()),
+        ];
+        let mut id = None;
+        for attribute in attribute_pk {
+            match attribute {
+                Attribute::Id(i) if id.is_none() => {
+                    template.push(Attribute::Id(i.clone()));
+                    id = Some(i);
+                }
+                _ => {}
+            }
+        }
+
+        let public_key = session.find_objects(&template)?.remove(0);
+        let attribute_pk = session.get_attributes(public_key, &[AttributeType::EcPoint])?;
+
+        let mut ec_point = None;
+        for attribute in attribute_pk {
+            match attribute {
+                Attribute::EcPoint(p) if ec_point.is_none() => {
+                    ec_point = Some(p);
+                }
+                _ => {}
+            }
+        }
+
+        let ec_point = ec_point.ok_or(Error::MissingAttribute(AttributeType::EcPoint))?;
+
+        // TODO(baloo): figure out why this is wrapped in an octet string
```

Comment:
Isn't this just the potential hybrid representation of X9.62?

Reply:
I'm not sure. This looks like an encoding issue in SoftHSMv2 to me.
https://github.com/opendnssec/SoftHSMv2/blob/develop/src/bin/util/softhsm2-util-ossl.cpp#L741

I don't think that follow the spec (https://www.rfc-editor.org/rfc/rfc5480#section-2.2, or https://safecurves.cr.yp.to/grouper.ieee.org/groups/1363/private/x9-62-09-20-98.pdf#page=22 (section 4.3.6 Point-to-Octet-String Conversion and section 6.2 Syntax for Finite Field Elements and Elliptic Curve Points))

I still need to confirm that.

Reply:
My build of softhsm uses botan, not openssl.

Reply:
botan 2.x.x which would have the exact same behavior: https://github.com/randombit/botan/blob/2.19.3/src/lib/pubkey/ec_group/point_gfp.cpp#L617

Reply:
nevermind, just me reading the spec wrong.